cpp/src/arrow/record_batch.cc

```diff
@@ -248,49 +248,101 @@ Result<std::shared_ptr<StructArray>> RecordBatch::ToStructArray() const {
                                        /*offset=*/0);
 }
 
+#define TYPE_CASE(type)                                                \
```

Comment:
Instead of using macros, please use visitor pattern:
```c++
template <typename Out>
struct ConvertColumnsToTensorVisitor {
  Out*& out_values;
  const ArrayData& in_data;

  template <typename T>
  Status Visit(const T&) {
    if constexpr (is_numeric(T::type_id)) {
      using In = typename T::c_type;
      auto in_values = in_data.GetSpan<In>(1);

      if constexpr (std::is_same_v<In, Out>) {
        memcpy(out_values, in_values.data(), in_values.size_bytes());
        out_values += in_values.size();
      } else {
        for (In in_value : in_values) {
          *out_values++ = static_cast<Out>(in_value);
        }
      }
      return Status::OK();
    }
    util::Unreachable();
  }
};
```

This would be invoked with VisitTypeInline:

```c++
  for (const auto& column : batch.columns()) {
    ConvertColumnsToTensorVisitor<CType> visitor{out_values, *column->data()};
    DCHECK_OK(VisitTypeInline(*column->type(), &visitor));
  }
```

Reply:
> Instead of using macros, please use visitor pattern:

Just to understand this better, is there a specific reason for this, or just a general project preference? (we use the visitor pattern a lot in the code base, but we _also_ have lots of places where we use this pattern with a macro)

Reply:
Macros tend to be brittle and as a matter of code style we advise not to use them when practical. There is also code written before the visitor pattern was available/well known. As this macro demonstrates we still haven't completed the job of normalizing visitor pattern in favor of type switches.

Reply:
Thank you for helping me out with the thorough example @bkietz ! Works great and looks super nice =)
https://github.com/apache/arrow/pull/40359/commits/b5cf637e8d0da49fb666f40459aac50a2907bfde